cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp

```diff
@@ -490,14 +495,132 @@ table_with_metadata hybrid_scan_reader_impl::materialize_payload_columns(
 
   auto output_dtypes = get_output_types(_output_buffers_template);
 
-  auto data_page_mask = _extended_metadata->compute_data_page_mask(
-    row_mask, row_group_indices, output_dtypes, _output_column_schemas, stream);
+  auto data_page_mask = _extended_metadata->compute_data_page_mask(row_mask,
+                                                                   row_group_indices,
+                                                                   output_dtypes,
+                                                                   _output_column_schemas,
+                                                                   _rows_processed_so_far,
+                                                                   stream);
 
-  prepare_data(row_group_indices, std::move(column_chunk_buffers), data_page_mask, options);
+  prepare_data(
+    read_mode::READ_ALL, row_group_indices, std::move(column_chunk_buffers), data_page_mask);
 
   return read_chunk_internal(read_mode::READ_ALL, read_columns_mode::PAYLOAD_COLUMNS, row_mask);
 }
 
+void hybrid_scan_reader_impl::setup_chunking_for_filter_columns(
+  std::size_t chunk_read_limit,
+  std::size_t pass_read_limit,
+  cudf::host_span<std::vector<size_type> const> row_group_indices,
+  cudf::host_span<std::vector<bool> const> page_mask,
+  std::vector<rmm::device_buffer> column_chunk_buffers,
+  parquet_reader_options const& options,
+  rmm::cuda_stream_view stream)
+{
+  CUDF_EXPECTS(not row_group_indices.empty(), "Empty input row group indices encountered");
+  CUDF_EXPECTS(options.get_filter().has_value(), "Encountered empty converted filter expression");
+
+  reset_internal_state();
+
+  table_metadata metadata;
+  populate_metadata(metadata);
+  _expr_conv = named_to_reference_converter(options.get_filter(), metadata);
+
+  CUDF_EXPECTS(_expr_conv.get_converted_expr().has_value(), "Filter expression must not be empty");
+
+  initialize_options(row_group_indices, options, stream);
+  _input_pass_read_limit   = pass_read_limit;
+  _output_chunk_read_limit = chunk_read_limit;
+
+  select_columns(read_columns_mode::FILTER_COLUMNS, options);
+
+  prepare_data(
+    read_mode::CHUNKED_READ, row_group_indices, std::move(column_chunk_buffers), page_mask);
+}
+
+table_with_metadata hybrid_scan_reader_impl::materialize_filter_columns_chunk(
+  cudf::mutable_column_view row_mask, rmm::cuda_stream_view stream)
+{
+  CUDF_EXPECTS(_file_preprocessed, "Chunking for filter columns not yet setup");
```

Comment:
Even the `materialize_xx_columns_chunk` look almost identical. Will be cleaned up